interface/main/main_screen.php

```diff
@@ -90,87 +89,162 @@ function generate_html_end()
+    echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "UTF") {
+        echo "<p style='color:red;font-weight:bold'>" . text($errormsg) . "</p>\n";
+    }
+
+    echo '<br /><h3>'.xl('or use').'</h3><br />';
+    echo '<h2>'.xl('TOTP Verification').'</h2>';
+    echo "<p>\n";
+    echo xlt('Enter the code from your authentication application on your device.');
+    echo "<table><tr>";
+    echo "<td><input type='text' name='totp' /></td><td>";
+    echo "<button type='submit'>".xla('Authenticate TOTP')."</button>\n";
+    echo "<input type='hidden' name='form_response' value='true' />\n";
+    echo "</td></tr></table>\n";
+    if ($errormsg && $errortype == "TOTP") {
+        echo "<span style='color:red;font-weight:bold'>" . text($errormsg) . "</span>\n";
+    }
+}
+
 
 $regs = array();          // for mapping device handles to their names
 $registrations = array(); // the array of stored registration objects
 $res1 = sqlStatement(
-    "SELECT a.name, a.var1 FROM login_mfa_registrations AS a " .
-    "WHERE a.user_id = ? AND a.method = 'U2F' ORDER BY a.name",
+    "SELECT a.name, a.method, a.var1 FROM login_mfa_registrations AS a " .
+    "WHERE a.user_id = ? ORDER BY a.name",
     array($_SESSION['authId'])
 );
 while ($row1 = sqlFetchArray($res1)) {
     $regobj = json_decode($row1['var1']);
     $regs[json_encode($regobj->keyHandle)] = $row1['name'];
     $registrations[] = $regobj;
 }
+
 if (!empty($registrations)) {
+    $requests = '';
+    $errortype = '';
     // There is at least one U2F key registered so we have to request or verify key data.
     // https is required, and with a proxy the server might not see it.
     $scheme = "https://"; // isset($_SERVER['HTTPS']) ? "https://" : "http://";
     $appId = $scheme . $_SERVER['HTTP_HOST'];
     $u2f = new u2flib_server\U2F($appId);
     $userid = $_SESSION['authId'];
     $form_response = empty($_POST['form_response']) ? '' : $_POST['form_response'];
+
     if ($form_response) {
-        // We have key data, check if it matches what was registered.
-        $tmprow = sqlQuery("SELECT login_work_area FROM users_secure WHERE id = ?", array($userid));
-        try {
-            $registration = $u2f->doAuthenticate(
-                json_decode($tmprow['login_work_area']), // these are the original challenge requests
-                $registrations,
-                json_decode($_POST['form_response'])
-            );
-            // Stored registration data needs to be updated because the usage count has changed.
-            // We have to use the matching registered key.
-            $strhandle = json_encode($registration->keyHandle);
-            if (isset($regs[$strhandle])) {
+        $res1 = sqlStatement(
+            "SELECT a.var1 FROM login_mfa_registrations AS a WHERE a.user_id = ? AND a.method = 'TOTP'",
+            array($_SESSION['authId'])
+        );
```

Comment:
TODO for me:
Seems odd doing this here when not sure if TOTP being used yet. Will be better to use sqlQuery here since just grabbing one item.